### ISSUE TITLE
Custom (and default?) headers

### ISSUE THREAD

**Issue by aseemk** [brian-gates/cypher-stream#10]
An important use case for us at @FiftyThree is to pass custom headers in the underlying requests. (And they vary across requests, so it's not just a static default.)

Examples of how we use those:

- **User-Agent**. The Neo4j team encourages drivers to set this (to `driver/version`), btw, so cypher-stream should do this by default I think (e.g. `cypher-stream/0.2.1`), but we're also thinking of expanding this to include our app's name, since we'll soon be having two apps talking to the same Neo4j database. (That'll let us differentiate and understand our load across the two.)

- **Tracing**. We send along a header with an ID for the outermost "app-level" request, so we can both visualize all the database requests that made up an app requests, and debug failed app requests. This is a common technique, e.g. [Twitter's Zipkin](http://twitter.github.io/zipkin/) (what we use), [Heroku's X-Request-ID](https://brandur.org/request-ids), etc.

- **Query names**. We want to be able to analyze the performance of our queries, but using the full query body would be untenable. So we send along a X-Query-Name header with every query that's a short, human-readable name for this query (e.g. `User_getByEmail`), which we then log. E.g.:

Example screenshot of the last case:

![home-stream-perf-queries](https://cloud.githubusercontent.com/assets/537838/4945779/53cd866a-660b-11e4-96bb-1e2de93c5bca.png)

So it'd be really valuable to be able to pass custom headers along to Cypher requests.

WDYT of supporting these alternate signature then?

```js
var cypher = require('cypher-stream')({
    url: 'http://localhost:7474',
    headers: {...}  // default headers, if you want to set any; e.g. User-Agent
});

cypher({
    query: 'match (user:User {email: {email}}) return user',
    params: {email: 'foo@example.com'},
    headers: {...}  // e.g. X-Query-Name
})
  .on('data', function (result){
    console.log(result.user.first_name);
  })
  .on('end', function() {
    console.log('all done');
  })
;
```

**Comment by aseemk** [comment]
Sorry for the delay here @brian-gates. I've finally started getting to this, but I'm realizing this isn't as simple as I first thought it'd be. (At least if I want this for transactional queries too, not just simple ones.)

In particular, because cypher-stream batches queries under the hood, there isn't a clean way to expose an API for query-specific headers (e.g. query name).

I realize the caller can set the batch size to 1 to achieve only one query per request, but the issue is the API. If I add `headers` onto `{statement, parameters}` objects, that doesn't make sense if the batch size *isn't 1*. But if I add `headers` onto a second `options` object (which would make more sense from an API perspective), it doesn't work with transactions (because `WritableStream::write` doesn't support an arbitrary second parameter; the parameters are `chunk, encoding, done` as you know).

The only solution I can think of is to add `headers` onto the `{statement, parameters}` object, and if it's present, immediately flush the batch queue and send the current query as its own request afterward. But that seems tacky and breaks the batching abstraction to me.

Any thoughts? Thanks!

**Comment by aseemk** [comment]
In pull #16, I went ahead and added support for `headers` within `{statement, parameters}` objects, because I saw you did something similar with `commit` and `rollback` as well.

For simplicity, I didn't bother implementing the "if present in a batch, immediately flush, etc.". That can be added later if it's needed. For now, it just uses the last `headers` value for the request, so it's up to the caller to ensure there's no batching (via `batchSize: 1`) if they want per-query headers.
